### DISCUSSION TITLE
How to update geometry in NiceGUI 3D scene?

### DISCUSSION THREAD

**Discussion by chengguizi** [zauberzeug/nicegui#1269] [Q&A]
### Question

Hi NiceGUI community,

Really greatful to find a streamlit alternative with more flexible control loops and seemingly lightweight runtime design concept.

I am looking at the capability to render 3D objects in NiceGUI. Documentation says it is based on threejs. Does that mean we can use all the functions within threejs?

But i see very few API reference here though https://nicegui.io/documentation/scene#inherited_from

What I am trying to achieve is like multi-line live update, based on streamed odometry data.

It looks something like this:
- https://threejs.org/docs/index.html#manual/en/introduction/How-to-update-things
- https://jsfiddle.net/t4m85pLr/1/

How could I achieve this in NIceGUI? Would be great to have some code example too. Thanks!

**Comment by chengguizi** [comment]
Also, would it possible with current API to visualise live pointcloud data in the 3d scene?

**Comment by falkoschindler** [comment]
I principle, you can update a 3D scene by either manipulating existing objects, or by replacing them like this:

```py
import numpy as np

scene = ui.scene()

def update_points(points: np.ndarray):
    # remove old points
    for obj in list(scene.objects.values()):
        if obj.name == 'point':
            obj.delete()

    # add new points
    with scene:
        for point in points:
            scene.sphere(0.1).material('red').move(*point).with_name('point')

ui.button('Update point cloud', on_click=lambda: update_points(np.random.randn(100, 3)))
```

In a similar way you can use `scene.line(...)` to create polylines.

Please note:
- While writing the example above I noticed a bug #1285. Currently the update is broken until you refresh the browser once.
- This method is not particularly efficient. Especially drawing spheres to represent tiny points is rather expensive. You can reduce `width_segments` and `height_segments` of a sphere to reduce the number of polygons. But a dedicated point cloud object would be much better.
- With #1213, which is scheduled for the next release, you won't necessarily need `with scene:` to create new objects.

**Comment by rodja** [comment]
I think we should add `scene.buffer_geometry` and `scene.pointcloud`? Anyone interested in taking this on?

**Comment by falkoschindler** [comment]
We just noticed that PR https://github.com/zauberzeug/nicegui/pull/545 already implements point clouds. 😄
We should add a point cloud demo.

**Reply by chengguizi** [reply]
Just wondering, would the same PR allow dynamically update the pointcloud object, after creation? The use case would be, to change the pose of the pointcloud due to the associated keyframe pose being changed.

**Reply by falkoschindler** [reply]
Sure:
```py
with ui.scene().classes('w-full h-64') as scene:
    x, y = np.meshgrid(np.linspace(-3, 3), np.linspace(-3, 3))
    z = np.sin(x) * np.cos(y) + 1
    points = np.dstack([x, y, z]).reshape(-1, 3)
    cloud = scene.point_cloud(points=points, colors=points, point_size=0.1)

ui.button('Move the points', on_click=lambda: cloud.move(x=1))
```
